grpc/src/main/java/io/stargate/grpc/service/Service.java

```diff
@@ -170,6 +209,174 @@ private void handleError(Throwable throwable, StreamObserver<?> responseObserver
     }
   }
 
+  private void handlePersistenceException(
+      PersistenceException pe, StreamObserver<?> responseObserver) {
+    switch (pe.code()) {
+      case SERVER_ERROR:
+      case PROTOCOL_ERROR: // Fallthrough
+      case UNPREPARED: // Fallthrough
+        onError(responseObserver, Status.INTERNAL, pe);
+        break;
+      case INVALID:
+      case SYNTAX_ERROR: // Fallthrough
+        onError(responseObserver, Status.INVALID_ARGUMENT, pe);
+        break;
+      case TRUNCATE_ERROR:
+      case CDC_WRITE_FAILURE: // Fallthrough
+        onError(responseObserver, Status.ABORTED, pe);
```

Comment:
What is the difference between `ABORTED` and `INTERNAL` is this case? It looks like we handle the lower-level error in a generic manner in this case and when we return `INTERNAL` so why is this client-side error code different here?

Reply:
Are we sure the mutation does not make any changes in this case?

Reply:
Here's the criteria I used for choosing different error codes:

> A litmus test that may help a service implementor in deciding between FAILED_PRECONDITION, ABORTED, and UNAVAILABLE: (a) Use UNAVAILABLE if the client can retry just the failing call. (b) Use ABORTED if the client should retry at a higher-level (e.g., restarting a read-modify-write sequence). (c) Use FAILED_PRECONDITION if the client should not retry until the system state has been explicitly fixed. E.g., if an "rmdir" fails because the directory is non-empty, FAILED_PRECONDITION should be returned since the client should not retry unless they have first fixed up the directory by deleting files from it.

> Why not expose more specific error codes to the client here?

These are gRPC specific status code. https://grpc.github.io/grpc-java/javadoc/io/grpc/Status.html

I guess we could have our own `Error` type in the result payload instead of using gRPC errors and `Status`, but I'd prefer to stay the gRPC way of handling errors unless there's some critical reason we shouldn't.

Reply:
Makes sense. Thanks for the info :+1: